database.go

```diff
@@ -0,0 +1,723 @@
+	idx := t.createMeasurementIfNotExists(name)
+
+	t.seriesToMeasurement[s.ID] = idx
+	t.series[s.ID] = s
+
+	// TODO: add this series to the global tag index
+
+	return idx.addSeries(s)
+}
+
+// createMeasurementIfNotExists will either add a measurement object to the index or return the existing one.
+func (t *database) createMeasurementIfNotExists(name string) *Measurement {
+	idx := t.measurements[name]
+	if idx == nil {
+		idx = NewMeasurement(name)
+		t.measurements[name] = idx
+		t.names = append(t.names, name)
+		sort.Strings(t.names)
+	}
+	return idx
+}
+
+// AddField adds a field to the measurement name. Returns false if already present
+func (t *database) AddField(name string, f *Field) bool {
+	panic("not implemented")
+	return false
+}
+
+// MeasurementsForSeriesIDs returns a collection of unique Measurements for the passed in SeriesIDs.
+func (t *database) MeasurementsForSeriesIDs(seriesIDs SeriesIDs) []*Measurement {
+	measurements := make(map[*Measurement]bool)
+
+	for _, id := range seriesIDs {
+		measurements[t.seriesToMeasurement[id]] = true
+	}
+
+	values := make([]*Measurement, 0, len(measurements))
+	for m, _ := range measurements {
+		values = append(values, m)
+	}
+
+	return values
+}
+
+// SeriesIDs returns an array of series ids for the given measurements and filters to be applied to all.
+// Filters are equivalent to and AND operation. If you want to do an OR, get the series IDs for one set,
+// then get the series IDs for another set and use the SeriesIDs.Union to combine the two.
+func (t *database) SeriesIDs(names []string, filters Filters) SeriesIDs {
+	// they want all ids if no filters are specified
+	if len(filters) == 0 {
+		ids := SeriesIDs(make([]uint32, 0))
+		for _, idx := range t.measurements {
+			ids = ids.Union(idx.ids)
+		}
+		return ids
+	}
+
+	ids := SeriesIDs(make([]uint32, 0))
+	for _, n := range names {
+		ids = ids.Union(t.seriesIDsForName(n, filters))
+	}
+
+	return ids
+}
+
+// TagKeys returns a sorted array of unique tag keys for the given measurements.
+func (t *database) TagKeys(names []string) []string {
+	if len(names) == 0 {
+		names = t.names
+	}
+
+	keys := make(map[string]bool)
+	for _, n := range names {
+		idx := t.measurements[n]
+		if idx != nil {
+			for k, _ := range idx.seriesByTagKeyValue {
+				keys[k] = true
+			}
+		}
+	}
+
+	sortedKeys := make([]string, 0, len(keys))
+	for k, _ := range keys {
+		sortedKeys = append(sortedKeys, k)
+	}
+	sort.Strings(sortedKeys)
+
+	return sortedKeys
+}
+
+// TagValues returns a map of unique tag values for the given measurements and key with the given filters applied.
+// Call .ToSlice() on the result to convert it into a sorted slice of strings.
+// Filters are equivalent to and AND operation. If you want to do an OR, get the tag values for one set,
+// then get the tag values for another set and do a union of the two.
+func (t *database) TagValues(names []string, key string, filters []*Filter) TagValues {
```

Comment:
This may just be me, but I find the variable `names` not that helpful. `measurements` might be better